File: tests/runtime/handles/host.ts
import { addImportsToImports, Imports } from "./imports.js";
import { Exports } from "./exports.js";
import * as exports from "./exports.js";
import { getWasm, addWasiToImports } from "./helpers.js";
// @ts-ignore
import * as assert from 'assert';

async function run() {
  const importObj = {};
  let sawClose = false;
  const imports: Imports = {
    hostStateCreate() { return 100; },
    hostStateGet(x) { return x as number; },
    hostState2Create() { return 101; },
    hostState2SawClose() { return sawClose; },
    dropHostState2(state) { sawClose = true; },
    twoHostStates(a, b) { return [b, a]; },
    hostState2ParamRecord(x) {},
    hostState2ParamTuple(x) {},
    hostState2ParamOption(x) {},
    hostState2ParamResult(x) {},
    hostState2ParamVariant(x) {},
    hostState2ParamList(x) {},

    hostState2ResultRecord() { return { a: {} }; },
    hostState2ResultTuple() { return [{}]; },
    hostState2ResultOption() { return 102; },
    hostState2ResultResult() { return { tag: 'ok', val: {} }; },
    hostState2ResultVariant() { return { tag: 0, val: {} }; },
    hostState2ResultList() { return [{}, 3]; },

    markdown2Create() {
      class Markdown {
        buf: string;

        constructor() {
          this.buf = '';
        }
        append(extra: string) {
          this.buf += extra;
        }
        render() {
          return this.buf.replace('red', 'green');
        }
      }

      return new Markdown();
    },

    oddNameCreate() {
      class OddName {
        frobTheOdd() {}
      }
      return new OddName();
    }
  };
  let instance: WebAssembly.Instance;
  addImportsToImports(importObj, imports, name => instance.exports[name]);
  const wasi = addWasiToImports(importObj);

  const wasm = new Exports();
  await wasm.instantiate(getWasm(), importObj);
  wasi.start(wasm.instance);
  instance = wasm.instance;

  wasm.testImports();

  // Param/result of a handle works in a simple fashion
  const s: exports.WasmState = wasm.wasmStateCreate();
  assert.strictEqual(wasm.wasmStateGetVal(s), 100);

  // Deterministic destruction is possible
  assert.strictEqual(wasm.wasmState2SawClose(), false);
  const s2: exports.WasmState2 = wasm.wasmState2Create();
  assert.strictEqual(wasm.wasmState2SawClose(), false);
  s2.drop();
  assert.strictEqual(wasm.wasmState2SawClose(), true);

  const arg1 = wasm.wasmStateCreate();
  const arg2 = wasm.wasmState2Create();
  const [c, d] = wasm.twoWasmStates(arg1, arg2);
  arg1.drop();
  arg2.drop();

  wasm.wasmState2ParamRecord({ a: d });
  wasm.wasmState2ParamTuple([d]);
  wasm.wasmState2ParamOption(d);
  wasm.wasmState2ParamOption(null);
  wasm.wasmState2ParamResult({ tag: 'ok', val: d });
  wasm.wasmState2ParamResult({ tag: 'err', val: 2 });
  wasm.wasmState2ParamVariant({ tag: 0, val: d });
  wasm.wasmState2ParamVariant({ tag: 1, val: 2 });
  wasm.wasmState2ParamList([]);
  wasm.wasmState2ParamList([d]);
  wasm.wasmState2ParamList([d, d]);

  c.drop();
  d.drop();

  wasm.wasmState2ResultRecord().a?.drop();
  wasm.wasmState2ResultTuple()[0].drop();
  const opt = wasm.wasmState2ResultOption();
  if (opt === null)
    throw new Error('should be some');
  opt.drop();
  const result = wasm.wasmState2ResultResult();
  if (result.tag === 'err')
    throw new Error('should be ok');
  result.val.drop();
  const variant = wasm.wasmState2ResultVariant();
  if (variant.tag === 1)
    throw new Error('should be 0');
  variant.val.drop();
  for (let val of wasm.wasmState2ResultList())
    val.drop();

  s.drop();

  const md = exports.Markdown.create(wasm);
  if (md) {
    md.append("red is the best color");
    assert.strictEqual(md.render(), "green is the best color");
    md.drop();
  }
}

await run()

File: tests/runtime/handles/host.py
from dataclasses import dataclass
from exports.bindings import Exports
from imports.bindings import add_imports_to_linker, Imports
from typing import Tuple, List
import exports.bindings as e
import imports.bindings as i
import sys
import wasmtime

@dataclass
class HostState(i.HostState):
    val: int

    def __init__(self, val: int) -> None:
        self.val = val

    def drop(self) -> None:
        pass


HOST_STATE2_CLOSED = False


@dataclass
class HostState2(i.HostState2):
    val: int

    def __init__(self, val: int) -> None:
        self.val = val

    def drop(self) -> None:
        global HOST_STATE2_CLOSED
        HOST_STATE2_CLOSED = True


@dataclass
class Markdown(i.Markdown2):
    buf: str = ''

    def append(self, data: str) -> None:
        self.buf += data

    def render(self) -> str:
        return self.buf.replace('red', 'green')

    def drop(self) -> None:
        pass


class OddName(i.OddName):
    def frob_the_odd(self) -> None:
        pass

    def drop(self) -> None:
        pass


class MyImports:
    def host_state_create(self) -> i.HostState:
        return HostState(100)

    def host_state_get(self, a: i.HostState) -> int:
        assert(isinstance(a, HostState))
        return a.val

    def host_state2_create(self) -> i.HostState2:
        return HostState2(101)

    def host_state2_saw_close(self) -> bool:
        return HOST_STATE2_CLOSED

    def two_host_states(self, a: i.HostState, b: i.HostState2) -> Tuple[i.HostState, i.HostState2]:
        return (b, a)

    def host_state2_param_record(self, a: i.HostStateParamRecord) -> None:
        pass

    def host_state2_param_tuple(self, a: i.HostStateParamTuple) -> None:
        pass

    def host_state2_param_option(self, a: i.HostStateParamOption) -> None:
        pass

    def host_state2_param_result(self, a: i.HostStateParamResult) -> None:
        pass

    def host_state2_param_variant(self, a: i.HostStateParamVariant) -> None:
        pass

    def host_state2_param_list(self, a: List[i.HostState2]) -> None:
        pass

    def host_state2_result_record(self) -> i.HostStateResultRecord:
        return i.HostStateResultRecord(HostState(2))

    def host_state2_result_tuple(self) -> i.HostStateResultTuple:
        return (HostState(2),)

    def host_state2_result_option(self) -> i.HostStateResultOption:
        return HostState(2)

    def host_state2_result_result(self) -> i.HostStateResultResult:
        return i.Ok(HostState2(2))

    def host_state2_result_variant(self) -> i.HostStateResultVariant:
        return HostState2(2)

    def host_state2_result_list(self) -> List[i.HostState2]:
        return [HostState2(2), HostState2(5)]

    def markdown2_create(self) -> i.Markdown2:
        return Markdown()

    def odd_name_create(self) -> i.OddName:
        return OddName()

def run(wasm_file: str) -> None:
    store = wasmtime.Store()
    module = wasmtime.Module.from_file(store.engine, wasm_file)
    linker = wasmtime.Linker(store.engine)
    linker.define_wasi()
    wasi = wasmtime.WasiConfig()
    wasi.inherit_stdout()
    wasi.inherit_stderr()
    store.set_wasi(wasi)

    imports = MyImports()
    add_imports_to_linker(linker, store, imports)
    wasm = Exports(store, linker, module)

    wasm.test_imports(store)

    # Param/result of a handle works in a simple fashion
    s: e.WasmState = wasm.wasm_state_create(store)
    assert(wasm.wasm_state_get_val(store, s) == 100)

    # Deterministic destruction is possible
    assert(wasm.wasm_state2_saw_close(store) == False)
    s2: e.WasmState2 = wasm.wasm_state2_create(store)
    assert(wasm.wasm_state2_saw_close(store) == False)
    s2.drop(store)
    assert(wasm.wasm_state2_saw_close(store) == True)

    arg1 = wasm.wasm_state_create(store)
    arg2 = wasm.wasm_state2_create(store)
    c, d = wasm.two_wasm_states(store, arg1, arg2)
    arg1.drop(store)
    arg2.drop(store)

    wasm.wasm_state2_param_record(store, e.WasmStateParamRecord(d))
    wasm.wasm_state2_param_tuple(store, (d,))
    wasm.wasm_state2_param_option(store, d)
    wasm.wasm_state2_param_option(store, None)
    wasm.wasm_state2_param_result(store, e.Ok(d))
    wasm.wasm_state2_param_result(store, e.Err(2))
    wasm.wasm_state2_param_variant(store, d)
    wasm.wasm_state2_param_variant(store, 2)
    wasm.wasm_state2_param_list(store, [])
    wasm.wasm_state2_param_list(store, [d])
    wasm.wasm_state2_param_list(store, [d, d])

    c.drop(store)
    d.drop(store)

    wasm.wasm_state2_result_record(store).a.drop(store)
    wasm.wasm_state2_result_tuple(store)[0].drop(store)
    opt = wasm.wasm_state2_result_option(store)
    assert(opt is not None)
    opt.drop(store)
    result = wasm.wasm_state2_result_result(store)
    assert(isinstance(result, e.Ok))
    result.value.drop(store)
    variant = wasm.wasm_state2_result_variant(store)
    print(variant)
    assert(isinstance(variant, e.WasmState2))
    variant.drop(store)
    for val in wasm.wasm_state2_result_list(store):
        val.drop(store)

    s.drop(store)

    md = e.Markdown.create(store, wasm)
    if md:
        md.append(store, "red is the best color")
        assert(md.render(store) == "green is the best color")
        md.drop(store)

if __name__ == '__main__':
    run(sys.argv[1])
